### PR TITLE
[lte][agw] Update CPE monitoring last reported time to milliseconds

### LINKED ISSUE #3735: Feature doc for enabling cpe monitoring
### Your Environment

- 1.3
- NMS; missing "last reported time"; CPE monitoring is not turned on on the AGWs.
- NMS and AGW
- AWS orc8r and BM AGW

### Describe the Issue
"Last Reported Time" in NMS is showing a default time for all subscribers in the nms. Looking a little more, it seems this value is not updated unless cpe monitoring is turned on. Currently a feature doc is required to enable cpe monitoring.

### PR DESCRIPTION
**Pull request by ardzoht**
## Summary

- `last_reported_time` on subscriber state is updated to float, to fix NMS dependency discrepancy (Date object expects float, due to this it's showing incorrect date)
- Fixing incorrect module naming for cpe_monitoring object
- Updating pipelined.yml to `10.1.0.1` to match mtr0 iface config
- Reformatting code on monitord module

## Test Plan

- Tested locally on AGW, and local Orc8r with eNB and UE to check ICMP monitoring reaches target and reports state correctly

```
(python) vagrant@magma-dev:~/magma/lte/gateway$ sudo journalctl -fu magma@monitord
-- Logs begin at Tue 2020-12-01 07:42:04 UTC. --
Dec 01 08:54:18 magma-dev monitord[13382]: ERROR:root:GetSubscribers Error for StatusCode.UNAVAILABLE! Connect Failed
Dec 01 08:54:18 magma-dev monitord[13382]: WARNING:root:No subscribers/ping targets found
Dec 01 08:55:31 magma-dev monitord[13382]: INFO:root:001010000000082:192.168.128.84 => 179.451ms
Dec 01 08:56:27 magma-dev monitord[13382]: INFO:root:Replicating state for 001010000000082: {"last_reported_time": 1606812931.343585, "latency_ms": 179.451}
Dec 01 08:56:45 magma-dev monitord[13382]: INFO:root:001010000000082:192.168.128.84 => 127.01ms
Dec 01 08:57:27 magma-dev monitord[13382]: INFO:root:Replicating state for 001010000000082: {"last_reported_time": 1606813005.555548, "latency_ms": 127.01}
Dec 01 08:57:58 magma-dev monitord[13382]: INFO:root:001010000000082:192.168.128.84 => 244.653ms
Dec 01 08:58:27 magma-dev monitord[13382]: INFO:root:Replicating state for 001010000000082: {"last_reported_time": 1606813078.78969, "latency_ms": 244.653}
Dec 01 08:59:12 magma-dev monitord[13382]: INFO:root:001010000000082:192.168.128.84 => 172.52ms
Dec 01 08:59:27 magma-dev monitord[13382]: INFO:root:Replicating state for 001010000000082: {"last_reported_time": 1606813152.992067, "latency_ms": 172.52}
Dec 01 09:00:27 magma-dev monitord[13382]: INFO:root:001010000000082:192.168.128.84 => 249.099ms
Dec 01 09:00:27 magma-dev monitord[13382]: INFO:root:Replicating state for 001010000000082: {"last_reported_time": 1606813227.191562, "latency_ms": 249.099}
Dec 01 09:01:27 magma-dev monitord[13382]: INFO:root:Replicating state for 001010000000082: {"last_reported_time": 1606813227.191562, "latency_ms": 249.099}
Dec 01 09:01:41 magma-dev monitord[13382]: INFO:root:001010000000082:192.168.128.84 => 152.87ms
Dec 01 09:02:28 magma-dev monitord[13382]: INFO:root:Replicating state for 001010000000082: {"last_reported_time": 1606813301.393722, "latency_ms": 152.87}
Dec 01 09:02:54 magma-dev monitord[13382]: INFO:root:001010000000082:192.168.128.84 => 239.474ms
Dec 01 09:03:28 magma-dev monitord[13382]: INFO:root:Replicating state for 001010000000082: {"last_reported_time": 1606813374.732781, "latency_ms": 239.474}
Dec 01 09:04:08 magma-dev monitord[13382]: INFO:root:001010000000082:192.168.128.84 => 183.536ms
Dec 01 09:04:28 magma-dev monitord[13382]: INFO:root:Replicating state for 001010000000082: {"last_reported_time": 1606813448.900638, "latency_ms": 183.536}
```

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/python/magma/monitord/cpe_monitoring.py [hunk 1]
```diff
@@ -25,73 +25,73 @@
...
-                             self._subscriber_state[sid].latency_ms))
-
-  def get_subscriber_state(self) -> Dict[str, ICMPMonitoringResponse]:
-      return self._subscriber_state
+PingedTargets = NamedTuple('PingedTargets',
+                           [('ping_targets', Dict['str', IPAddress]),
+                            ('ping_addresses', List[IPAddress])])
 
 
+def _get_addr_from_subscribers(sub_ip) -> str:
```
Comment (karthiksubraveti) [inline lte/gateway/python/magma/monitord/cpe_monitoring.py:40]: is just formatting change ?

File: lte/gateway/python/magma/monitord/main.py [hunk 2]
```diff
@@ -32,23 +32,27 @@ def main():
...
+    icmp_monitor = ICMPMonitoring(cpe_monitor,
+                                  service.mconfig.polling_interval,
                                   service.loop, mtr_interface)
     icmp_monitor.start()
 
     # Register a callback function for GetOperationalStates
     service.register_operational_states_callback(
         lambda: serialize_subscriber_states(
-            icmp_monitor.get_subscriber_state()))
+            cpe_monitor.get_subscriber_state()))
```
Comment (karthiksubraveti) [inline lte/gateway/python/magma/monitord/main.py]: why did this change ?

**Review by karthiksubraveti — COMMENTED** [review]
@ardzoht Do we have a unit test to verify this feature ?

File: lte/gateway/python/magma/monitord/cpe_monitoring.py (hunk 1 above)
Comment (ardzoht) [reply lte/gateway/python/magma/monitord/cpe_monitoring.py:40]: correct

File: lte/gateway/python/magma/monitord/main.py (hunk 2 above)
Comment (ardzoht) [reply lte/gateway/python/magma/monitord/main.py]: `IcmpMonitoring` module doesn't contain `get_subscriber_state`, is contained under `CpeMonitoringModule`

**Comment by ardzoht** [comment]
> @ardzoht Do we have a unit test to verify this feature ?

We have some basic unit tests under monitord module, it might be a good idea to expand on these, I can look into that in a separate issue.

**Review by karthiksubraveti — APPROVED** [review]

File: lte/gateway/configs/pipelined.yml [hunk 3]
```diff
@@ -131,7 +131,7 @@ redis_enabled: false
 magma_print_grpc_payload: false
 
 # MTR iface IP
-mtr_ip: 10.0.2.10
+mtr_ip: 10.1.0.1
```
Comment (pshelar) [inline lte/gateway/configs/pipelined.yml:134]: this is already set in gtp_br0 interface config, do we still need this?

File: lte/gateway/python/magma/monitord/cpe_monitoring.py [hunk 4]
```diff
@@ -25,73 +25,73 @@
...
+        self._subscriber_state = defaultdict(ICMPMonitoringResponse)
+        self.ping_addresses = []
+        self.ping_targets = {}
+
+    def set_manually_configured_targets(self,
+                                        configured_ping_targets:
+                                        Optional[Dict] = None):
+        if configured_ping_targets:
+            self.ping_targets = configured_ping_targets.copy()
+            for value in self.ping_targets.values():
```
Comment (pshelar) [inline lte/gateway/python/magma/monitord/cpe_monitoring.py:58]: lets plan to use icmplib once we move to python 3.8.

**Review by pshelar — APPROVED** [review]

File: lte/gateway/configs/pipelined.yml (hunk 3 above)
Comment (ardzoht) [reply lte/gateway/configs/pipelined.yml:134]: I think we use this for the ARP flows, @koolzz  might have more context on this

File: lte/cloud/go/services/subscriberdb/obsidian/handlers/handlers_test.go [hunk 5]
```diff
@@ -434,7 +434,7 @@ func TestListSubscribers(t *testing.T) {
 				ActiveApns: subscriberModels.ApnList{apn2, apn1},
 				Monitoring: &subscriberModels.SubscriberStatus{
 					Icmp: &subscriberModels.IcmpStatus{
-						LastReportedTime: frozenClock,
+						LastReportedTime: frozenClock * 1000,
```
Comment (hcgatewood) [inline lte/cloud/go/services/subscriberdb/obsidian/handlers/handlers_test.go:438]: Here and below: add short, inline comment explaining the `*1000`

File: lte/cloud/go/services/subscriberdb/obsidian/models/swagger.v1.yml [hunk 6]
```diff
@@ -505,6 +505,7 @@ definitions:
       last_reported_time:
```
Comment (hcgatewood) [inline lte/cloud/go/services/subscriberdb/obsidian/models/swagger.v1.yml:505]: If possible please add description field to `last_reported_time` property (idk if Swagger will let you?)

File: lte/gateway/python/magma/monitord/cpe_monitoring.py [hunk 7]
```diff
@@ -25,73 +25,73 @@
...
-
-  def get_subscriber_state(self) -> Dict[str, ICMPMonitoringResponse]:
-      return self._subscriber_state
+PingedTargets = NamedTuple('PingedTargets',
+                           [('ping_targets', Dict['str', IPAddress]),
+                            ('ping_addresses', List[IPAddress])])
 
 
+def _get_addr_from_subscribers(sub_ip) -> str:
+    return str(ipaddress.IPv4Address(sub_ip.address)
```
Comment (hcgatewood) [inline lte/gateway/python/magma/monitord/cpe_monitoring.py]: To me, a multi-line ternary return is less readable than just using an if-then

File: lte/gateway/python/magma/monitord/cpe_monitoring.py [hunk 8]
```diff
@@ -25,73 +25,73 @@
...
+
+
+class CpeMonitoringModule:
+    def __init__(self):
+        # TODO: Save to redis
+        self._subscriber_state = defaultdict(ICMPMonitoringResponse)
+        self.ping_addresses = []
+        self.ping_targets = {}
+
+    def set_manually_configured_targets(self,
```
Comment (hcgatewood) [inline lte/gateway/python/magma/monitord/cpe_monitoring.py]: Simlar to above, this style of line continuation is, to me, inherently less readable. Prefer the golang-style when possible, something like

```
def set_manually_configured_targets(
    self,
    configured_ping_targets: Optional[Dict] = None,
):
```

File: lte/gateway/python/magma/monitord/cpe_monitoring.py [hunk 9]
```diff
@@ -25,73 +25,73 @@
 from magma.monitord.icmp_state import ICMPMonitoringResponse
 from orc8r.protos.common_pb2 import Void
 from prometheus_client import Histogram
-from magma.subscriberdb.sid import SIDUtils
 
 SUBSCRIBER_ICMP_LATENCY_MS = Histogram('subscriber_icmp_latency_ms',
```
Comment (hcgatewood) [inline lte/gateway/python/magma/monitord/cpe_monitoring.py]: To me, ALL_CAPS should be reserved for const+immutable. This variable is not immutable, so I prefer its capitalization to be lower_snake_case

File: lte/gateway/python/magma/monitord/icmp_monitoring.py [hunk 10]
```diff
@@ -79,13 +76,9 @@ def _save_ping_response(self, target_id: str, ip_addr: str,
             self._module.save_ping_response(target_id, ip_addr, ping_resp)
 
     async def _run(self) -> None:
```
Comment (hcgatewood) [inline lte/gateway/python/magma/monitord/icmp_monitoring.py:78]: For my understanding: does this still run in a loop? It looks like you removed the explicit loop

File: lte/gateway/python/magma/monitord/main.py [hunk 11]
```diff
@@ -32,23 +32,27 @@ def main():
...
+    cpe_monitor.set_manually_configured_targets(manual_ping_targets)
 
-    icmp_monitor = ICMPMonitoring(obj, service.mconfig.polling_interval,
+    icmp_monitor = ICMPMonitoring(cpe_monitor,
+                                  service.mconfig.polling_interval,
                                   service.loop, mtr_interface)
     icmp_monitor.start()
 
     # Register a callback function for GetOperationalStates
     service.register_operational_states_callback(
```
Comment (hcgatewood) [inline lte/gateway/python/magma/monitord/main.py:58]: Personally I don't like using lambdas. Clearer+cleaner to use a named function

**Review by hcgatewood — APPROVED** [review]
Cloud changes lgtm, couple small notes -- please address cloud comments before merging

**Pull request merged**